README
Folder Color © 2012-2014 ﻿﻿Marcos Alvarez Costales
================================================

WHAT IS IT?
===========
A Nautilus extension for choosing the color of a folder in Ubuntu with just a right-click.



HOW DO I INSTALL & RUN IT?
==========================
From a PPA:
    # add-apt-repository ppa:costales/folder-color
    # apt-get update && sudo apt-get install folder-color
From the code (take a look to the dependencies):
    $ bzr branch lp:folder-color && cd folder-color && sudo python setup.py install --prefix=/usr

Then, restart Nautilus
    $ nautilus -q
You may also need to refresh the system icons:
    # gtk-update-icon-cache /usr/share/icons/hicolor/



DEPENDENCIES
============
For Nautilus:
    python-nautilus, nautilus, libgtk2.0-bin
For Nemo:
    python-nemo, nemo, libgtk2.0-bin
For Caja:
    python-caja, caja, gir1.2-caja, libgtk2.0-bin



WHAT DOES IT LOOK LIKE?
=======================
Take a look here!
http://foldercolor.tuxfamily.org/




WHAT IS NEW?
============
  * Ligther version of RAVEfinity icons



LICENSE
====================
Folder Color code is licensed under the GPL v3.
http://www.gnu.org/licenses



ICONS LICENSE
=============
The icons are licensed under GPL v2. Folder Color is mixing 2 themes:
- RAVEfinity artwork http://www.ravefinity.com/
- GNOME artwork http://commons.wikimedia.org/wiki/GNOME_Desktop_icons





HOW TO CREATE A THEME
=====================
1. THE TRICK
Folder Color is reading the icons from /usr/share/icons/hicolor by default.
If you create the same icons in your theme path, they will overwrite the icons in Folder Color. Then, create your icons under this directory:
    /usr/share/icons/<your_theme>


2. CREATE A QUICK & BASIC THEME
2.1: SPECIAL CUSTOM COLOR FILE
Folder Color will try to find this important file:
    /usr/share/icons/<your_theme>/scalable/places/folder_color_custom.svg
This file has to exists in your theme always! This icon is not using colors, it's using 3 constants inside the file. Then, when the user choose a custom color, Folder Color is replacing those constants as:
    #value_light = Color selected by user in hexadecimal.
    #value_middle = (Color selected by user in hexadecimal - 2500)
    #value_dark = (Color selected by user in hexadecimal - 5000)

2.2: COLORS
    /usr/share/icons/<your_theme>/scalable/places/folder_color_black.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_brown.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_blue.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_green.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_grey.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_orange.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_pink.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_red.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_purple.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_yellow.svg

2.3: EMBLEMS
    /usr/share/icons/<your_theme>/scalable/emblems/folder_emblem_favorite.svg
    /usr/share/icons/<your_theme>/scalable/emblems/folder_emblem_finished.svg
    /usr/share/icons/<your_theme>/scalable/emblems/folder_emblem_important.svg
    /usr/share/icons/<your_theme>/scalable/emblems/folder_emblem_in_progress.svg

2.4: CONTEXTUAL MENU ENTRIES:
    /usr/share/icons/<your_theme>/16x16/actions/folder_color_picker.svg
    /usr/share/icons/<your_theme>/16x16/actions/folder_color_edit.svg
    /usr/share/icons/<your_theme>/16x16/actions/folder_color_undo.svg


3. COMPLETE THEME (OPTIONAL)
3.1: PRECONFIGURED USER FOLDERS
The default user folders have different emblems: Desktop, Documents, Download, Music, Pictures, Public, Templates and Videos.
Then, for each color, create these icons (this example is just for the blue color):
    /usr/share/icons/<your_theme>/scalable/places/folder_color_blue_desktop.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_blue_documents.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_blue_downloads.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_blue_music.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_blue_pictures.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_blue_public.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_blue_templates.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_blue_videos.svg


4. FULL THEME (OPTIONAL)
4.1: CUSTOM COLORS FOR DEFAULT FOLDERS:
Custom color for Default folders too (Read the previously 2.1 section about these custom files):
    /usr/share/icons/<your_theme>/scalable/places/folder_color_custom_desktop.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_custom_documents.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_custom_downloads.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_custom_music.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_custom_pictures.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_custom_public.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_custom_templates.svg
    /usr/share/icons/<your_theme>/scalable/places/folder_color_custom_videos.svg

4.2: RESOLUTIONS
Reply the icons for all resolutions:
    16x16/places/folder_color_<color>_<default_folder>.svg
    22x22/places/folder_color_<color>_<default_folder>.svg
    24x24/places/folder_color_<color>_<default_folder>.svg
    32x32/places/folder_color_<color>_<default_folder>.svg
    48x48/places/folder_color_<color>_<default_folder>.svg
    64x64/places/folder_color_<color>_<default_folder>.svg
    128x128/places/folder_color_<color>_<default_folder>.svg

    
5. HOW TO CHECK?
Copy your icons into
    /usr/share/icons/<your_theme>/
After install them, remember to refresh your theme:
    # gtk-update-icon-cache /usr/share/icons/<your_theme>
And restart Nautilus:
    $ nautilus -q


6. WHAT ICONS TO DO? MY PERSONAL RECOMMENDATION
Make all the icons is a really big work.
I'd recommend you to create just the icons 2.1 & 2.2. If you want more: 2.3 & 2.4.
If you are full of beans, create 3.1 too.
With the time, if a lot of users are using this extension, extend it with the 4.1 & 4.2.


7. DOUBTS?
Do you have any doubt or problem? Contact me here: https://launchpad.net/~costales/+contactuser I'll try to help you :)






PREINSTALL FOLDER COLOR IN YOUR OS
==================================
If you want to include Folder Color preinstalled in a distro, I allow you to remove the Donation menu entry (I understand it has no place for you).
How could you hide the Donation menu entry by default? Create this folder for all users (/etc/skel directory):
~/.config/folder-color/hide_donation
I'll not receive donations from the users of your OS, then, if you can, a small donation from your distro will be really appreciate!


debian/changelog
folder-color (0.0.46) trusty; urgency=low

  * Ligther version of RAVEfinity icons

 -- costales <costales.marcos@gmail.com>  Sun, 7 Dec 2014 10:10:31 +0000



icons/converter.py
import os
import shutil

addons = [
['folder',''],
['folder-documents','_documents'],
['folder-downloads','_downloads'],
['folder-music','_music'],
['folder-pictures','_pictures'],
['folder-publicshare','_public'],
['folder-templates','_templates'],
['folder-videos','_videos'],
['user-desktop','_desktop']
]

colors = [
['blue', '#2980b9', '#3498db', '#3c2f1b'],
['brown', '#6A220B', '#B64926', '#3c2f1b'],
['custom', '#2c3e50', '#34495e', '#ecf0f1'],
['green', '#27ae60', '#2ecc71', '#3c2f1b'],
['grey', '#95a5a6', '#bdc3c7', '#3c2f1b'],
['orange', '#d35400', '#e67e22', '#3c2f1b'],
['pink', '#EA2E49', '#FF4B66', '#3c2f1b'],
['purple', '#8e44ad', '#9b59b6', '#3c2f1b'],
['red', '#c0392b', '#e74c3c', '#3c2f1b'],
['yellow', '#f39c12', '#f1c40f', '#3c2f1b']
]

emblems = [
['emblem-ok','folder_emblem_finished'],
['emblem-favorite','folder_emblem_favorite'],
['emblem-important','folder_emblem_important'],
['emblem-synchronizing','folder_emblem_in_progress'],
]

folders = [
'16x16',
'22x22',
'24x24',
'32x32',
'48x48',
'64x64',
'128x128',
'256x256',
'scalable'
]

if not os.path.isdir('hicolor'):
	os.mkdir('hicolor')

		
for folder in folders:

	if not os.path.isdir('hicolor/' + folder):
		os.mkdir('hicolor/' + folder)

	if not os.path.isdir('hicolor/' + folder + '/places'):
		os.mkdir('hicolor/' + folder + '/places')

	if not os.path.isdir('hicolor/' + folder + '/emblems'):
		os.mkdir('hicolor/' + folder + '/emblems')


	directory = '/usr/share/icons/Numix/' + folder + '/places'

	for addon in addons:
		svg = directory + '/' + addon[0] + '.svg';

		if not os.path.isfile(svg):
			svg = '/usr/share/icons/Numix/256x256/places' + '/' + addon[0] + '.svg';

		f = open(svg, 'r')
		data = f.read()
		f.close()

		for color in colors:
			name = color[0]
			color1 = color[1]
			color2 = color[2]
			color3 = color[3]

			changed = data.replace('#e9a439', color1)
			changed = changed.replace('#f5c14e', color2)
			changed = changed.replace('#3c2f1b', color3)

			f = open('hicolor/' + folder + '/places/folder_color_' + name + addon[1] + '.svg', 'w')
			f.write(changed)
			f.close()


	directory = '/usr/share/icons/Numix/' + folder + '/emblems'

	for emblem in emblems:
		svg = directory + '/' + emblem[0] + '.svg';
		if not os.path.isfile(svg):
			svg = '/usr/share/icons/Numix/256x256/emblems' + '/' + emblem[0] + '.svg';

		f = open(svg, 'r')
		data = f.read()
		f.close()

		f = open('hicolor/' + folder + '/emblems/' + emblem[1] + '.svg', 'w')
		f.write(data)
		f.close()







install_scripts/caja.sh
#!/bin/bash
rm -r ../icons
rm -r ../po
rm ../README

# setup
sed -i '25,32d' ../setup.py
sed -i 's/]),/])]/' ../setup.py
sed -i 's/Change your folder color with just a click/Change your folder color in Caja/' ../setup.py
sed -i 's/nautilus-python/caja-python/' ../setup.py
sed -i 's/nautilus-extension/caja-extension/' ../setup.py
sed -i 's/"folder-color"/"folder-color-caja"/' ../setup.py

# extension
mv ../nautilus-extension/ ../caja-extension
sed -i 's/nautilus/caja/g' ../caja-extension/folder-color.py
sed -i 's/Nautilus/Caja/g' ../caja-extension/folder-color.py
sed -i 's/\"metadata::custom-icon-name\", new_color)/\"metadata::custom-icon\", \"\".join([\"file:\/\/\/usr\/share\/icons\/hicolor\/scalable\/places\/\", new_color, \".svg\"]\)\)/' ../caja-extension/folder-color.py

# debian
rm ../debian/postinst

sed -i '2d' ../debian/install
sed -i 's/nautilus/caja/g' ../debian/install

sed -i 's/Upstream-Name: folder-color/Upstream-Name: folder-color-caja/' ../debian/copyright
sed -i '25,45d' ../debian/copyright

sed -i 's/Source: folder-color/Source: folder-color-caja/' ../debian/control
sed -i 's/Package: folder-color/Package: folder-color-caja/' ../debian/control
sed -i 's/python-nautilus, nautilus, libgtk2.0-bin, /python-caja, gir1.2-caja, caja, gir1.2-gtk-2.0, folder-color-common, /' ../debian/control
sed -i '14,15d' ../debian/control
sed -i 's/Folder Color for Nautilus/Folder Color for Caja/' ../debian/control
sed -i 's/Change a folder color used in Nautilus/Change a folder color used in Caja/' ../debian/control

sed -i 's/folder-color/folder-color-caja/' ../debian/changelog

# me
rm -r ../install_scripts



install_scripts/nemo.sh
#!/bin/bash
rm -r ../icons
rm -r ../po
rm ../README

# setup
sed -i '25,32d' ../setup.py
sed -i 's/]),/])]/' ../setup.py
sed -i 's/Change your folder color with just a click/Change your folder color in Nemo/' ../setup.py
sed -i 's/nautilus-python/nemo-python/' ../setup.py
sed -i 's/nautilus-extension/nemo-extension/' ../setup.py
sed -i 's/"folder-color"/"folder-color-nemo"/' ../setup.py

# extension
mv ../nautilus-extension/ ../nemo-extension
sed -i 's/nautilus/nemo/g' ../nemo-extension/folder-color.py
sed -i 's/Nautilus/Nemo/g' ../nemo-extension/folder-color.py

# debian
rm ../debian/postinst

sed -i '2d' ../debian/install
sed -i 's/nautilus/nemo/g' ../debian/install

sed -i 's/Upstream-Name: folder-color/Upstream-Name: folder-color-nemo/' ../debian/copyright
sed -i '25,45d' ../debian/copyright

sed -i 's/Source: folder-color/Source: folder-color-nemo/' ../debian/control
sed -i 's/Package: folder-color/Package: folder-color-nemo/' ../debian/control
sed -i 's/python-nautilus, nautilus, libgtk2.0-bin, /python-nemo, nemo, folder-color-common, /' ../debian/control
sed -i '14,15d' ../debian/control
sed -i 's/Folder Color for Nautilus/Folder Color for Nemo/' ../debian/control
sed -i 's/Change a folder color used in Nautilus/Change a folder color used in Nemo/' ../debian/control

sed -i 's/folder-color/folder-color-nemo/' ../debian/changelog

# me
rm -r ../install_scripts



install_scripts/common.sh
#!/bin/bash
rm -r ../nautilus-extension
rm ../README

# setup
sed -i 's/data = .*/data = [/' ../setup.py
sed -i 's/"folder-color"/"folder-color-common"/' ../setup.py
sed -i 's/Change your folder color with just a click/Folder Color Library/' ../setup.py

# po
sed -i 's/folder-color/folder-color-common/' ../po/POTFILES.in
sed -i '4d' ../po/POTFILES.in

# debian
sed -i '1d' ../debian/install
sed -i 's/Upstream-Name: folder-color/Upstream-Name: folder-color-common/' ../debian/copyright
sed -i 's/Source: folder-color/Source: folder-color-common/' ../debian/control
sed -i 's/Package: folder-color/Package: folder-color-common/' ../debian/control
sed -i 's/python-nautilus, nautilus, /gvfs-bin, /' ../debian/control
sed -i 's/Breaks: .*/Breaks: folder-color (<< 0.0.37), folder-color-nemo (<< 0.0.37), folder-color-caja (<< 0.0.37)/' ../debian/control
sed -i 's/Replaces: .*/Replaces: folder-color (<< 0.0.37), folder-color-nemo (<< 0.0.37), folder-color-caja (<< 0.0.37)/' ../debian/control
sed -i 's/Folder Color for Nautilus/Folder Color Library/' ../debian/control
sed -i 's/Change a folder color used in Nautilus/Icons and translations for Folder Color/' ../debian/control
sed -i 's/folder-color/folder-color-common/' ../debian/changelog

# me
rm -r ../install_scripts


nautilus-extension/folder-color.py
# -*- coding: utf-8 -*-
# Folder Color 0.0.46
# Copyright (C) 2012-2014 Marcos Alvarez Costales https://launchpad.net/~costales
#
# Folder Color is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
# 
# Folder Color is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with Folder Color; if not, see http://www.gnu.org/licenses
# for more information.

import os, urllib, gettext, webbrowser, glob, subprocess
from gi.repository import Gtk, Nautilus, GObject, Gio, GLib

# i18n
gettext.textdomain('folder-color-common')
_ = gettext.gettext


class ChangeColorFolder(GObject.GObject, Nautilus.MenuProvider):
    def __init__(self):
        dummy_default_i18n = (_("Brown"), _("Blue"), _("Green"), _("Grey"), _("Orange"), _("Pink"), _("Red"), _("Purple"), _("Yellow"), _("Favorite"), _("Finished"), _("Important"), _("In Progress")) # Need i18n, because colors/emblems are dynamic now
        if not len(self.get_menu_items()):
            return # No icons = Nothing to do
        
        self.GRADIENT_MIDDLE_RANGE = 2500
        self.GRADIENT_DARK_RANGE = 5000
        self.HIDE_DONATION = os.path.join(os.getenv('HOME'), '.config', 'folder-color', 'hide_donation')
        self.FOLDER_ICONS_CUSTOM = os.path.join(os.getenv('HOME'), '.config', 'folder-color', 'custom_icons')
        self.DEFAULT_FOLDERS = self.get_default_folders()
        
        if not os.path.exists(self.FOLDER_ICONS_CUSTOM):
            try:
                os.makedirs(self.FOLDER_ICONS_CUSTOM)
            except OSError as exception:
                pass
            except:
                pass
        pass
    
    def get_theme_path(self, what='places'):
        default_theme_path = '/usr/share/icons/hicolor/scalable' # hardcode, because if not colors, not entry by init
        
        icon_theme = Gtk.IconTheme.get_default()
        icon_info = icon_theme.lookup_icon('folder_color_custom', 0, 0)
        try:
            theme_path = icon_info.get_filename().replace('/places/folder_color_custom.svg', '') # Will be something like /usr/share/icons/Theme/scalable/places/folder_color_custom.svg
            # Hack for emblems: Exist theme, but without emblems, get hicolor emblems
            if what == 'emblem':
                emblems_in_theme = glob.glob(os.path.join(theme_path, 'emblems/folder_emblem_*.svg'))
                if not len(emblems_in_theme):
                    theme_path = default_theme_path
        except:
            theme_path = default_theme_path
        
        if what == 'emblem':
            return os.path.join(theme_path, 'emblems')
        else:
            return os.path.join(theme_path, 'places')
    
    def get_menu_items(self, get_type='color'):
        available_items = {}
        available_icons = glob.glob(os.path.join(self.get_theme_path(get_type), 'folder_' + get_type + '_*.svg'))
        for each_icon in available_icons:
            each_icon = os.path.basename(each_icon)
            if not any(not_count in each_icon for not_count in ('undo', 'edit', 'picker', 'custom', 'desktop', 'documents', 'downloads', 'music', 'pictures', 'public', 'templates', 'videos')):
                color = each_icon.replace('folder_' + get_type + '_', '').replace('.svg', '').replace('_', ' ')
                available_items[color] = _(color.title())
        return available_items
    
    def get_default_folders(self):
        folders = {}
        folders[GLib.get_user_special_dir(GLib.USER_DIRECTORY_DESKTOP)]      = 'desktop'
        folders[GLib.get_user_special_dir(GLib.USER_DIRECTORY_DOCUMENTS)]    = 'documents'
        folders[GLib.get_user_special_dir(GLib.USER_DIRECTORY_DOWNLOAD)]     = 'downloads'
        folders[GLib.get_user_special_dir(GLib.USER_DIRECTORY_MUSIC)]        = 'music'
        folders[GLib.get_user_special_dir(GLib.USER_DIRECTORY_PICTURES)]     = 'pictures'
        folders[GLib.get_user_special_dir(GLib.USER_DIRECTORY_PUBLIC_SHARE)] = 'public'
        folders[GLib.get_user_special_dir(GLib.USER_DIRECTORY_TEMPLATES)]    = 'templates'
        folders[GLib.get_user_special_dir(GLib.USER_DIRECTORY_VIDEOS)]       = 'videos'
        return folders
    
    def get_icon_name(self, folder, icon, get_type='color'):
        icon = icon.replace(' ', '_') # Come from the file name
        
        # Available preconfigured icon?
        try:
            if os.path.isfile(os.path.join(self.get_theme_path(get_type), 'folder_%s_%s_%s.svg' % (get_type, icon, self.DEFAULT_FOLDERS[folder]))):
                return '_'.join(['folder', 'color', icon, self.DEFAULT_FOLDERS[folder]])
        except:
            pass
        
        # Default icon
        return '_'.join(['folder', get_type, icon])
    
    def choose_custom_color(self):
        colorselectiondialog = Gtk.ColorSelectionDialog(_("Choose a Folder Color"))
        selector = colorselectiondialog.get_color_selection()
        response = colorselectiondialog.run()
        colorselectiondialog.destroy()
        
        if response == Gtk.ResponseType.OK:
            color = selector.get_current_color()
            # Control max color range
            if color.red <= self.GRADIENT_DARK_RANGE:
                color.red = self.GRADIENT_DARK_RANGE
            if color.green <= self.GRADIENT_DARK_RANGE:
                color.green = self.GRADIENT_DARK_RANGE
            if color.blue <= self.GRADIENT_DARK_RANGE:
                color.blue = self.GRADIENT_DARK_RANGE
            # Convert to hex
            rgb_ligth = (color.red, color.green, color.blue)
            rgb_middle = (color.red-self.GRADIENT_MIDDLE_RANGE, color.green-self.GRADIENT_MIDDLE_RANGE, color.blue-self.GRADIENT_MIDDLE_RANGE)
            rgb_dark = (color.red-self.GRADIENT_DARK_RANGE, color.green-self.GRADIENT_DARK_RANGE, color.blue-self.GRADIENT_DARK_RANGE)
            htmlcolor_light =  ''.join((str(hex(i/257))[2:].rjust(2, '0') for i in rgb_ligth)).upper()
            htmlcolor_middle = ''.join((str(hex(i/257))[2:].rjust(2, '0') for i in rgb_middle)).upper()
            htmlcolor_dark =   ''.join((str(hex(i/257))[2:].rjust(2, '0') for i in rgb_dark)).upper()
            return (htmlcolor_light, htmlcolor_middle, htmlcolor_dark)
        
        return None
    
    def generate_icon_from_template(self, folder, colors):
        # Parsing template to a custom color
        orig_file = os.path.join(self.get_theme_path(), self.get_icon_name(folder, 'custom') + '.svg')
        dest_file = os.path.join(self.FOLDER_ICONS_CUSTOM, ''.join([self.get_icon_name(folder, 'custom'), '_', colors[0], '.svg']))
        
        f_input = open(orig_file, 'r') # This has to exists always
        f_output = open(dest_file, 'w')
        
        f_generate  = f_input.read().replace('value_light', colors[0]).replace('value_middle', colors[1]).replace('value_dark', colors[2])
        f_output.write(f_generate)
        
        return dest_file
    
    def reload_icon(self, folder):
        # Reload the current folder
        os.system('touch "%s"' % folder)
    
    def menu_activate_color(self, menu, color, folders):
        # Custom
        if color == 'custom':
            custom_color = self.choose_custom_color()
            # Cancel?
            if custom_color == None:
                return
        
        for each_folder in folders:
            if each_folder.is_gone():
                continue
            
            # Get info
            folder_name = urllib.unquote(each_folder.get_uri()[7:])
            folder = Gio.File.new_for_path(folder_name)
            info = folder.query_info('metadata::custom-icon-name', 0, None)
            
            # Restore always previously, because I don't know if I set previously with custom-icon or custom-icon-name
            info.set_attribute('metadata::custom-icon', Gio.FileAttributeType.INVALID, 0)
            info.set_attribute('metadata::custom-icon-name', Gio.FileAttributeType.INVALID, 0)
            
            # Set color
            if not color == 'restore':
                if not color == 'custom':
                    new_color = self.get_icon_name(folder_name, color)
                    info.set_attribute_string("metadata::custom-icon-name", new_color)
                else:
                    file_custom_color = ''.join(['file://', self.generate_icon_from_template(folder_name, custom_color)]) # Generate file color
                    info.set_attribute_string('metadata::custom-icon', file_custom_color)
            else:
                self.set_emblem(folder_name, '') # Restore all
            
            # Write changes
            folder.set_attributes_from_info(info, 0, None)
            
            # Reload changes
            self.reload_icon(folder_name)
    
    def set_emblem(self, folder, emblem=''):
        # TODO: Use Gio for doing this
        if emblem:
            subprocess.call('gvfs-set-attribute -t stringv "%s" metadata::emblems %s' % (folder, emblem), shell=True)
        else:
            subprocess.call('gvfs-set-attribute -t unset "%s" metadata::emblems' % folder, shell=True)
            
    def menu_activate_emblem(self, menu, emblem, folders):
        for each_folder in folders:
            if each_folder.is_gone():
                continue
            
            folder = urllib.unquote(each_folder.get_uri()[7:])
            new_emblem = ''
            if not emblem == 'restore':
                new_emblem = ''.join(['folder_emblem_', emblem.replace(' ', '_')])
            
            self.set_emblem(folder, new_emblem)
            
            # Reload changes
            self.reload_icon(folder)
    
    def menu_activate_donate(self, menu):
        webbrowser.open_new('http://gufw.org/donate_foldercolor')
        # Hide donation after show it
        try:
            os.makedirs(self.HIDE_DONATION)
        except OSError as exception:
            pass
        except:
            pass
        return
    
    # Nautilus invoke this function in its startup > Then, create menu entry
    def get_file_items(self, window, items_selected):
        if not len(self.get_menu_items()):
            return # No icons = Nothing to do
        
        # No items selected
        if not len(items_selected):
            return
        for each_item in items_selected:
            # GNOME can only handle files
            if each_item.get_uri_scheme() != 'file':
                return
            # Only folders
            if not each_item.is_directory():
                return
        
        # Main menu [1 or +1 folder(s)]
        if len(items_selected) > 1:
            top_menuitem = Nautilus.MenuItem(name='ChangeFolderColorMenu::Top', label=_("Folders' Color"), icon='folder_color_picker')
        else:
            top_menuitem = Nautilus.MenuItem(name='ChangeFolderColorMenu::Top', label=_("Folder's Color"),  icon='folder_color_picker')
        submenu = Nautilus.Menu()
        top_menuitem.set_submenu(submenu)
        
        # Colors submenu
        sorted_colors = sorted(self.get_menu_items().items(), key=lambda x:x[1])
        for color in sorted_colors:
            name = ''.join(['ChangeFolderColorMenu::Colors"', color[0], '"'])
            label = color[1]
            item = Nautilus.MenuItem(name=name, label=label, icon=self.get_icon_name('Default', color[0].lower()))
            item.connect('activate', self.menu_activate_color, color[0], items_selected)
            submenu.append_item(item)
        
        # Custom color
        item_custom = Nautilus.MenuItem(name=''.join(['ChangeFolderColorMenu::"custom"']), label=_("Custom"), icon='folder_color_edit')
        item_custom.connect('activate', self.menu_activate_color, 'custom', items_selected)
        submenu.append_item(item_custom)
        
        # Restore
        # Separator
        item_sep = Nautilus.MenuItem(name='ChangeFolderEmblemMenu::Sep', label=_("Restore"), sensitive=False)
        submenu.append_item(item_sep)
        
        item_restore = Nautilus.MenuItem(name='ChangeFolderColorMenu::Restore', label=_("By Default"), icon='folder_color_undo')
        item_restore.connect('activate', self.menu_activate_color, 'restore', items_selected)
        submenu.append_item(item_restore)
        
        # Emblems submenu
        sorted_emblems = sorted(self.get_menu_items('emblem').items(), key=lambda x:x[1])
        for i,emblem in enumerate(sorted_emblems):
            # Separator
            if not i:
                item_sep = Nautilus.MenuItem(name='ChangeFolderColorMenu::Sep1', label=_("Categorize"), sensitive=False)
                submenu.append_item(item_sep)
            
            name = ''.join(['ChangeFolderEmblemMenu::Emblem"', emblem[0], '"'])
            label = emblem[1]
            item = Nautilus.MenuItem(name=name, label=label, icon=self.get_icon_name('Default', emblem[0].lower(), 'emblem'))
            item.connect('activate', self.menu_activate_emblem, emblem[0], items_selected)
            submenu.append_item(item)
        
        # Donation
        if not os.path.exists(self.HIDE_DONATION):
            # Separator
            item_sep = Nautilus.MenuItem(name='ChangeFolderColorMenu::Sep2', label=_("Support this app"), sensitive=False)
            submenu.append_item(item_sep)
            # Donation
            item = Nautilus.MenuItem(name=''.join(['ChangeFolderColorMenu::"donate"']), label=_("Donate? Click for hiding"))
            item.connect('activate', self.menu_activate_donate)
            submenu.append_item(item)
        
        return top_menuitem,
